exercises/symbolic_sequences_data/descr.md

```diff
@@ -0,0 +1,177 @@
+of the sequence in turn. The complexity of `get` and `foreach` is a little
+more difficult to describe; let us say that, if a sequence is built without
+using `map`, then the complexity of `get` is linear in the depth of the
+sequence description, and the complexity of `foreach` is linear in the length
+of the sequence plus the size of the sequence description.
+
+**In this exercise, we implement symbolic sequences as data**, that is,
+as a (generalized) algebraic data type. In a companion
+exercise, entitled "Symbolic Sequences as Objects", we implement them as
+records of functions. Although these two approaches give
+rise to two different ways of organizing the code, the underlying algorithm is
+the same. In fact, the GADT-based code is the image of the object-based code
+through a mechanical transformation known as
+[defunctionalization](https://en.wikipedia.org/wiki/Defunctionalization).
+
+In another exercise, entitled "Enumerating Trees",
+we *use* symbolic sequences.
+
+## Interface
+
+Before explaining how symbolic sequences are implemented,
+let us first describe their interface,
+that is,
+the operations that they support.
+
+`'a seq` is the type of a sequence whose elements have type `'a`.
+```
+  type 'a seq
+```
+
+### Constructors
+
+There are five constructor functions:
+`empty`,
+`singleton`,
+`sum`,
+`product`,
+and `map`.
+
+The constant `empty` is the empty sequence.
+
+```
+  val empty: 'a seq
+```
+
+The sequence `singleton x` has length 1. Its single element is `x`.
+
+```
+  val singleton: 'a -> 'a seq
+```
+
+The sequence `sum s1 s2` is the concatenation of the sequences `s1` and `s2`.
+
+```
+  val sum: 'a seq -> 'a seq -> 'a seq
+```
+
+The sequence `product s1 s2`, a sequence of pairs, is the Cartesian product
+of the sequences `s1` and `s2`. Its length is the product of the lengths of
+`s1` and `s2`. The first pair component is considered most significant,
+which means that the product of the integer sequence `0; 1` by itself is the
+sequence `(0, 0); (0, 1); (1, 0); (1, 1)`.
+
+```
+  val product: 'a seq -> 'b seq -> ('a * 'b) seq
+```
+
+The sequence `map phi s` is the image of the sequence `s` through the function
+`phi`.
+
+```
+  val map: ('a -> 'b) -> 'a seq -> 'b seq
+```
+
+### Destructors
+
+There are three destructors: `length`, `get`, and `foreach`.
+
+`length s` returns the length of the sequence `s`.
+
+```
+  val length: 'a seq -> int
+```
+
+If the index `i` is comprised between zero (included) and `length s` (excluded),
+then `get s i` returns the `i`-th element of the sequence `s`.
+Otherwise, `get s i` raises the exception `OutOfBounds`.
+
+```
+  exception OutOfBounds
+  val get: 'a seq -> int -> 'a
+```
+
+`foreach s k` iterates over all elements of the sequence `s`.
+Each element in turn is passed to the loop body `k`.
+Thus, the OCaml code `foreach s (fun x -> ...)` can be
+informally understood as `for each x in s do ...`.
+
+```
+  val foreach: 'a seq -> ('a -> unit) -> unit
+```
```

Comment:
I guess this type could be generalized to:
```
val foreach : 'a seq -> 'b -> ('b -> 'a -> 'b) -> 'b
```
which allows for purely functional iteration on the list contrary to its `'b = unit` instantiation. 

I guess you did not want to pollute the code with accumulator passing but it may be a good idea to mention somewhere (as a new question) that `fold` is also feasible.

Reply:
Indeed, I wanted to keep things simple. The note at the end of the assignment already suggests how an `elements` function can be implemented using `foreach`. I don't think there would be much value in adding another question or remark about `fold`. (Maybe I am just feeling lazy...)